app/src/main/java/com/gh4a/widget/ReactionBar.java

```diff
@@ -474,110 +481,99 @@ public AddReactionMenuHelper(@NonNull Context context, Menu menu,
+                    mReactionMenuItems.add(reactionMenuItem);
+                    Drawable icon = DrawableCompat.wrap(reactionMenuItem.getIcon().mutate());
+                    DrawableCompat.setTintMode(icon, PorterDuff.Mode.SRC_ATOP);
+                    reactionMenuItem.setIcon(icon);
+                }
             }
         }
 
-        public boolean onItemClick(MenuItem item) {
-            for (int i = 0; i < mItems.length; i++) {
-                if (item == mItems[i]) {
-                    item.setChecked(mOldReactionIds[i] == 0);
-                    addOrRemoveReaction(CONTENTS[i], mOldReactionIds[i]);
-                    updateDrawableState();
+        public boolean onItemClick(MenuItem clickedItem) {
+            for (MenuItem item : mReactionMenuItems) {
+                if (clickedItem == item) {
+                    String reactionContent = getReactionContentByViewId(item.getItemId());
+                    Long userReactionId = mUserOwnReactions.get(reactionContent);
+                    addOrRemoveReaction(reactionContent, userReactionId);
                     return true;
                 }
             }
             return false;
         }
 
-        public void update() {
-            List<Reaction> reactions = mDetailsCache.getEntry(mItem);
-            if (reactions != null) {
-                mOldReactionIds = new long[mItems.length];
-                mLastKnownDetails = new ArrayList<>(reactions);
-
-                String ownLogin = Gh4Application.get().getAuthLogin();
-                for (Reaction reaction : reactions) {
-                    if (!ApiHelpers.loginEquals(reaction.user(), ownLogin)) {
-                        continue;
-                    }
-                    for (int i = 0; i < CONTENTS.length; i++) {
-                        if (TextUtils.equals(CONTENTS[i], reaction.content())) {
-                            mOldReactionIds[i] = reaction.id();
-                            break;
-                        }
-                    }
-                }
-            } else {
-                mOldReactionIds = null;
-                mLastKnownDetails = null;
+        public void updateMenuItems() {
+            mUserOwnReactions.clear();
+            List<Reaction> reactions = mDetailsCache.getReactions(mItem);
+            if (reactions == null) {
+                setReactionMenuItemsVisible(false);
+                return;
             }
-            setDataItemsVisible(mOldReactionIds != null);
+
+            String ownLogin = Gh4Application.get().getAuthLogin();
+            reactions.stream()
+                    .filter(reaction -> ApiHelpers.loginEquals(reaction.user(), ownLogin))
+                    .forEach(reaction -> mUserOwnReactions.put(reaction.content(), reaction.id()));
+            setReactionMenuItemsVisible(true);
         }
 
         public void startLoadingIfNeeded() {
-            if (mOldReactionIds != null) {
-                syncCheckStates();
```

Comment:
It turned out that this check here was not needed at all, I'm not even sure why it was added (maybe to avoid re-running the entire logic to update menu items in some situations?)
I believe that it could have also caused the popup menu to display incorrect data when `AddReactionMenuHelper` is used in `CommentViewHolder`, since the same instance of the view holder and the menu helper is reused between different comments and therefore the data stored in `mOldReactionIds` could have referred to a different comment than the one bound to the view holder at a certain moment.